src/nodes/state_manager.rs

```diff
@@ -0,0 +1,800 @@
+                    "Tried to remove invalid negative index from cache: {}",
+                    at_index
+                );
+                return;
+            } else if at_index == -1 {
+                self.cached_states.len() - 1
+            } else {
+                at_index as usize
+            }
+        };
+        if usize_index >= self.cached_states.len() {
+            godot_error!(
+                "Specified cache index out of bounds; index is {} but cache length is {}!",
+                usize_index,
+                self.cached_states.len()
+            );
+            return;
+        }
+        self.cached_states.remove(usize_index as usize);
+    }
+
+    #[func]
+    // A cached state has a tag associated with it; this tag can be any variant type.
+    // Tags are not unique, but are the only way for a user to associate some custom item with their cached state.
+    fn cache_state_with_tag(&mut self, in_space: Rid, tag: Variant) {
+        if let Some(export_state) = self.fetch_state(in_space) {
+            let cached_state = CachedState {
+                state: RawImportState::from_export_state(export_state),
+                tag,
+            };
+            self.push_state_to_cache(cached_state);
+        }
+    }
+
+    #[func]
+    // A variation with no tag.
+    fn cache_state(&mut self, in_space: Rid) {
+        if let Some(export_state) = self.fetch_state(in_space) {
+            let cached_state = CachedState {
+                state: RawImportState::from_export_state(export_state),
+                tag: Variant::nil(),
+            };
+            self.push_state_to_cache(cached_state);
+        }
+    }
+
+    #[func]
+    // Loads the nth state from the cache. -1 grabs last entry. Only the subset of the cache with a matching tag will be considered.
+    fn load_cached_state_with_tag(
+        &mut self,
+        in_space: Rid,
+        mut index: i32,
+        with_tag: Variant,
+    ) -> bool {
+        // If no tag was provided, consider our subset to be all our cached states.
+        let subset: Vec<&CachedState> = {
+            if with_tag.is_nil() {
+                self.cached_states.iter().collect()
+            } else {
+                // Otherwise, if a tag was provided, our subset is the cache members that match the tag.
+                self.cached_states
+                    .iter()
+                    .filter(|c| c.tag == with_tag)
+                    .collect()
+            }
+        };
+        if subset.is_empty() {
+            return false;
+        };
+        if index == -1 {
+            index = (subset.len() as i32) - 1;
+        }
+        if let Some(cached_state) = subset.get(index as usize) {
+            self.load_state_internal(in_space, cached_state.state.clone());
+            return true;
+        }
+        false
+    }
+
+    #[func]
+    // Variation with no tag.
+    fn load_cached_state(&mut self, in_space: Rid, mut index: i32) -> bool {
```

Comment:
Since we offer the cache mechanism, I would not offer this function at all, since ppl can just call load_cached_state(..., null) (without tag)

Also, is there function to get_state_from_cache ? (probl is but the names of functions are a bit missleading). Also, what is the index here, if this function has an index, I'm assuming it's loading from an array of possible entries?
I would call this if such: load_cached_state_from_index(..)

And the other one load_cached_state_from_tag(..)

I this the API is a bit complicated right now. I would offer just one variant of it, either with tag or with index, but having multiple might make it harder to read/understand.